pkg/operator/hostendpointscontroller/host_endpoints_controller.go

```diff
@@ -0,0 +1,342 @@
+	machineConfigInformers := dynamicInformers.ForResource(schema.GroupVersionResource{
+		Group: "machineconfiguration.openshift.io", Version: "v1", Resource: "controllerconfigs"})
+
+	c := &HostEndpointsController{
+		eventRecorder: eventRecorder.WithComponentSuffix("host-etcd-endpoints-controller"),
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "HostEndpointsController"),
+		cachesToSync: []cache.InformerSynced{
+			operatorClient.Informer().HasSynced,
+			endpointsInformer.Informer().HasSynced,
+			podInformer.Informer().HasSynced,
+			machineConfigInformers.Informer().HasSynced,
+		},
+		operatorClient:      operatorClient,
+		machineConfigLister: machineConfigInformers.Lister(),
+		podLister:           podInformer.Lister(),
+		endpointsLister:     endpointsInformer.Lister(),
+		endpointsClient:     kubeClient.CoreV1(),
+	}
+	operatorClient.Informer().AddEventHandler(c.eventHandler())
+	endpointsInformer.Informer().AddEventHandler(c.eventHandler())
+	machineConfigInformers.Informer().AddEventHandler(c.eventHandler())
+	podInformer.Informer().AddEventHandler(c.eventHandler())
+	return c
+}
+
+func (c *HostEndpointsController) sync() error {
+	err := c.syncHostEndpoints()
+
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "HostEndpointsDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorUpdatingHostEndpoints",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			c.eventRecorder.Warning("HostEndpointsErrorUpdatingStatus", updateErr.Error())
+			return updateErr
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "HostEndpointsDegraded",
+		Status: operatorv1.ConditionFalse,
+		Reason: "HostEndpointsUpdated",
+	}))
+	if updateErr != nil {
+		c.eventRecorder.Warning("HostEndpointsErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *HostEndpointsController) syncHostEndpoints() error {
+	discoveryDomain, err := c.getEtcdDiscoveryDomain()
+	if err != nil {
+		return fmt.Errorf("unable to determine etcd discovery domain: %v", err)
+	}
+
+	// list etcd member pods
+	etcdPodRequirement, err := labels.NewRequirement("k8s-app", selection.In, []string{"etcd"})
+	if err != nil {
+		return err
+	}
+	etcdPodSelector := labels.NewSelector().Add(*etcdPodRequirement)
+	pods, err := c.podLister.List(etcdPodSelector)
+
+	// get dns names of ready etc member pods
+	var addresses []string
+	for _, pod := range pods {
+		var ready bool
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == corev1.PodReady {
+				ready = condition.Status == corev1.ConditionTrue
+				break
+			}
+		}
+		if ready {
+			dnsName, err := c.getEtcdDNSName(discoveryDomain, pod.Status.HostIP)
```

Comment:
with the possibility of transient DNS failures should we consider a retry here? If we error would that result in setting operator Degraded? Since we are within the service network this lookup is against the internal resolver right? The actual DNS lookups for this by kube-aspiserver would happen on the host network DNS resolver. I am looking for ways to mitigate DNS flakes that would set operator degraded when it is actually operating fine.